File: java/linkedlist/reverselinkedlist/ReverseLinkedList.java
package linkedlist.reverselinkedlist;

import java.util.Scanner;

public class ReverseLinkedList {

    public Node reverse(Node head){

        if(head == null || head.next == null){
            return head;
        }

        
        Node current = head;
        //Reverse the list here.
        Node prev = null;
        while(current != null){
            Node next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        return prev;
    }
    
    public static void main(String[] args) {
        
        ReverseLinkedList reverse = new ReverseLinkedList();
        System.out.println("Enter the number of elements you wish to add in the linkedlist.");
        try(Scanner scan = new Scanner(System.in)){
            int num = scan.nextInt();
            if(num <= 0){
                System.out.println("You should have atleast 1 node in the list.");
                return;
            }
            int i = 1;
            System.out.println("Enter the first element:");
            int firstItem = scan.nextInt();
            Node head = new Node(firstItem);
            Node current = head;

            while(i < num){
                //Building the linkedlist to pass to reverse method.
                System.out.println("Enter next element:");
                int next = scan.nextInt();
                Node node = new Node(next);
                current.next = node;
                current = node;
                i++;
            }

            Node reversedList = reverse.reverse(head);
            Node pointer = reversedList;
            while(pointer != null){
                System.out.print(" " + pointer.data + " -> ");
                pointer = pointer.next;
            }
        }
    }
}

class Node{
     int data;
     Node next;

     public Node(int data){
        this.data = data;
     }
}

File: kotlin/linkedlist/reverselinkedlist/ReverseLinkedList.kt
fun reverseLinkedList(node: Node): Node?{
    if (node == null) return node
    var current: Node? = node
    var prev: Node? = null
    while (current != null){
        var nextNode: Node? = current.next
        current.next = prev
        prev = current
        current = nextNode
    }

    return prev
}

fun display(node: Node?): Unit{
    if (node == null) return
    var current: Node? = node
    while (current != null){
        print("${current.data} ")
        current = current.next
    }
}

fun main() {
    println("Reverse a linked list soltion.")
    //Create a linked list
    val head: Node = Node(1)
    var node1 = Node(2)
    head.next = node1
    var node2 = Node(3)
    node1.next = node2
    var node3 = Node(4)
    node2.next = node3
    //reverse the linked list
    val node: Node? = reverseLinkedList(head)
    display(node)
}

class Node(var data: Int){
    var next: Node? = null

}